text/3579-pki.md

```diff
@@ -0,0 +1,262 @@
+
+We leave the implementation of these solutions for a future RFC - however, this provides us with the first steps in having the available cryptographic keys to begin exploring and discussing solutions to these problems.
+
+## Private Repositories
+
+The technical details of this proposal involve the initial signing of the index and crate files utilizing the Rust Project root key for validation. However, the implementation of this in cargo will allow for private repositories to validate against additional certificates.  This means all new security guarantees made with the new Rust keys can also be leveraged by private repositories for private certificate chains. (Cargo will not, however, allow the substitution of different keys for `crates.io` itself.)
+
+## Provenance
+
+Utilizing our PKI and certificate hierarchy, end users (manually or utilizing `rustup` and `cargo`) will be able to cryptographically verify the provenance of arbitrary files, regardless of source, as being from the Rust Project.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We propose leveraging [AWS CloudHSM Quorum Authentication](https://docs.aws.amazon.com/cloudhsm/latest/userguide/quorum-authentication.html) for the management of the `Root Key`. See [Alternatives#Quorum] for out-of-band quorum models which were considered.
+
+All operations using the root key or subkeys will be logged transparently, and will make use of Certificate Transparency / Binary Transparency infrastructure to ensure that no surreptitious signing operations can take place.
+
+## Keys and Trust
+
+**Key Algorithm:** ECDSA secp384r1
+
+**Storage:** PKCS #11 Certificates Stored in CloudHSM
+
+**Expiration:** Root and Top-Level Delegate keys shall follow a `7 year expiration` schedule, except for the first keys of this proposal, which shall have an expiration date of the expected release date of the Rust 2030 edition plus 1 year.
+
+### First-Issue Expiration Plan
+
+We shall schedule expiration on the key to match the `Rust Edition Cycle`, so any key expiration will also fall within the time frame of a new Rust edition. This cadence is chosen to ease the rotation and release and dissemination of a new `Root Key` aligning with an edition and thus, new releases of `cargo` and associated tooling.
+
+The first key shall be created with an expiration of `2031`, or 8 years, allowing for rotation of the key with the `2030 Edition` of Rust.
+
+All subsequent keys shall be created with a `7 year` expiration, aligning with the date of 2 Rust editions into the future, with 1 additional year allowed for key deployment prior to rotation.
+
+- First Root Key: Generated 2024, 7 year expiration. Expires: 2031, Rotated with: `2030 Edition`
+- Second Root Key: Generated 2030, 7 year expiration. Expires: 2037, Rotated with `2036 Edition`
+- This schedule shall continue until the end of the universe.
```

Comment:
We've good odds that future legislation raises this 7 years across many products, maybe not open source compiler tool chains of course, but this sort of short expiry period sreates considerable ewaste in phones, etc.

Reply:
@burdges We're about to push an update that addresses how people who need to run more-than-7-year-old toolchains can handle that case. (They can either update root keys or disable signature authentication, whichever works for them. The latter may make sense if they end up vendoring all their dependencies anyway and never touching the network.)